### PR TITLE
XLS-33d: Compact Fungible Tokens (CFTs)

### PR DESCRIPTION
**Pull request by ledhed2222**
Initial commit of XLS-33d spec

Per `CONTRIBUTING.md` - this is ready to merge as a draft spec. The spec will continue to be updated with new PRs after this!

### CODE REVIEW

**Review by sappenin — CHANGES_REQUESTED** [review]

File: XLS-33d/README.md [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+| `LedgerEntryType`   | :heavy_check_mark: | `number`  | `UINT16`      |
+| `Flags`             | :heavy_check_mark: | `number`  | `UINT32`      |
+| `Issuer`            | :heavy_check_mark: | `string`  | `ACCOUNTID`   |
+| `AssetCode`         | :heavy_check_mark: | `string`  | `UINT160`     |
+| `AssetScale`        | (default)          | `number`  | `UINT8`       |
+| `MaximumAmount`     | :heavy_check_mark: | `string`  | `UINT64`      |
+| `OutstandingAmount` | :heavy_check_mark: | `string`  | `UINT64`      |
+| `LockedAmount`      | ️(default)          | `string`  | `UINT64`      |
+| `TransferFee`       | ️(default)          | `number`  | `UINT16`      |
+| `Metadata`          |                    | `string`  | `BLOB`        |
```
Comment (sappenin) [inline XLS-33d/README.md:92]:
```suggestion
| `CFTMetadata`          |                    | `string`  | `BLOB`        |
```

I think we'll need to change this name so it doesn't collide with `sfMetadata` which is a `METADATA` type that we can't (or don't want to) use for CFT (in the rippled codebase, a `BLOB` is actually an `STI_VL` type).

File: XLS-33d/README.md [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+ {
+     "LedgerEntryType": "CFTokenIssuance",
+     "Flags" : 131072,
+     "Issuer" : "rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW",
+     "AssetCode": "5553440000000000000000000000000000000000",
+     "AssetScale": "2",
+     "MaximumAmount", "100000000",
+     "OutstandingAmount", "5",
+     "TransferFee": 50000,     
+     "Metadata" : "",
```
Comment (sappenin) [inline XLS-33d/README.md:156]: Per the above comment, this should be `CFTMetadata` (assuming the JSON property name must align with the SField definition).

File: XLS-33d/README.md [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+
+| Field Name      | Required?          | JSON Type | Internal Type |
+| --------------- | ------------------ | --------- | ------------- |
+| `MaximumAmount` | :heavy_check_mark: | `string`  | `UINT64`      | 
+
+The maximum asset amount of this token that should ever be issued.
+
+| Field Name | Required?          | JSON Type | Internal Type |
+| ---------- | ------------------ | --------- | ------------- |
+| `Metadata` | :heavy_check_mark: | `string`  | `BLOB`        |
```
Comment (sappenin) [inline XLS-33d/README.md:395]:
```suggestion
| `CFTMetadata ` | :heavy_check_mark: | `string`  | `BLOB`        | 
```

File: XLS-33d/README.md [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+
+```
+{
+  "TransactionType": "CFTokenIssuanceCreate",
+  "AssetCode": "5553440000000000000000000000000000000000",
+  "AssetScale": "2",
+  "TransferFee": 314,
+  "MaxAmount": "50000000",
+  "Flags": 83659,
+  "Metadata": "FOO",
```
Comment (sappenin) [inline XLS-33d/README.md:409]:
```suggestion
  "CFTMetadata": "FOO",
```

**Review by sappenin — APPROVED** [review]
This LGTM as a `DRAFT`.

File: XLS-0033d-CFTs/README.md [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+
+# 1. Compact Fungible Tokens (CFTs)
+
+## 1.1. Abstract
+
+[Trustlines](https://xrpl.org/trust-lines-and-issuing.html#trust-lines-and-issuing) are a fundamental building block for a variety of XRP Ledger tokenization features, including [CBDCs](https://en.wikipedia.org/wiki/Central_bank_digital_currency) and [fiat-collateralized stablecoins](https://www.investopedia.com/terms/s/stablecoin.asp). However, as more and more token issuers embrace the XRP Ledger, the current size of each trustline will become an impediment to ledger stability and accessibility.
+
+This proposal introduces extensions to the XRP Ledger to support a more compact fungible token (i.e., CFT) type, along with operations to enumerate, purchase, sell and hold such tokens. 
+
+Unlike Trustlines, CFTs do not represent bidirectional debt relationships. Instead, CFTs function more like a unidirectional trustline with only one balance, making it simpler to support common tokenization requirements, including even non-monetery use-cases such as tracking reputation points in an online game.
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:20]:
```suggestion
Unlike trust lines, CFTs do not represent bidirectional debt relationships. Instead, CFTs function more like a unidirectional trust line with only one balance, making it simpler to support common tokenization requirements, including even non-monetary use cases such as tracking reputation points in an online game. 
```

File: XLS-0033d-CFTs/README.md [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+
+Perhaps as important, however, CFTs require significantly less space than trustlines: ~52 bytes for each CFT held by a token holder, as compared to _at least_ 234 bytes for every new trustline (see Appendix 1 for a more detailed comparison).
+
+### 1.1.1. Advantages and Disadvantages
+
+**Advantages**
+
+* Uses a fixed-point balance representation instead of a floating-point representation, yielding the following benefits:
+  * CFT balance amounts can easily be added to other ledger objects like escrows, checks, payment channels, and AMMs.
+  * The implemenatation can more reliably and easily enforce invariant checks and explicitly track fees.
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:30]:
```suggestion
  * Enables reliable and easy enforcement of invariant checks and straightforward tracking of fees.
```

File: XLS-0033d-CFTs/README.md [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+  * Eliminates edge case floating-point math involving very small amounts violating expected equality conditions (e.g., CFTs will never have to deal with cases where `A+B=A` for non-zero `B` or `(A+B)+C != A+(B+C)`). 
+* Simpler conceptual model (truslines and rippling make it harder for developers to reason about the system, which increases the risk of errors or value loss).
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:32]:
```suggestion
* Simpler conceptual model (truslines and rippling make it harder for developers to reason about the system, which increases the risk of errors or value loss).
```
```suggestion
* Simpler conceptual model (trust lines and rippling make it harder for developers to reason about the system, which increases the risk of errors or value loss).
```

File: XLS-0033d-CFTs/README.md [hunk 8]
```diff
@@ -0,0 +1,836 @@
...
+| `LedgerEntryType`   | :heavy_check_mark: | `number`  | `UINT16`      |
+| `Flags`             | :heavy_check_mark: | `number`  | `UINT32`      |
+| `Issuer`            | :heavy_check_mark: | `string`  | `ACCOUNTID`   |
+| `AssetCode`         | :heavy_check_mark: | `string`  | `UINT160`     |
+| `AssetScale`        | (default)          | `number`  | `UINT8`       |
+| `MaximumAmount`     | :heavy_check_mark: | `string`  | `UINT64`      |
+| `OutstandingAmount` | :heavy_check_mark: | `string`  | `UINT64`      |
+| `LockedAmount`      | ️(default)          | `string`  | `UINT64`      |
+| `TransferFee`       | ️(default)          | `number`  | `UINT16`      |
+| `CFTMetadata`          |                    | `string`  | `BLOB`        |
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:92]:
```suggestion
| `TransferFee`       | ️(default)          | `number`  | `UINT16`      |
| `CFTMetadata`       |                    | `string`  | `BLOB`        |
```

File: XLS-0033d-CFTs/README.md [hunk 9]
```diff
@@ -0,0 +1,836 @@
...
+
+###### 1.2.1.1.2.3. `Issuer`
+
+The address of the account that controls both the issuance amounts and characteristics of a particular fungible token.
+
+###### 1.2.1.1.2.4. `AssetCode`
+
+A 160-bit blob of data. We recommend using only upper-case ASCII letters, ASCII digits 0 through 9, the dot (`.`) character, and the dash (`-`) character. Dots and dashes should never be the first character of an asset code, and should not be repeated sequentially.
+
+While it's possible to store any arbitrary data in this field, implementations that detect the above recommended character conformance can and should display them as ASCII for human readability, allowing issuers to support well-known ISO-4207 currency codes in addition to custom codes. This also helps prevents spoofing attacks where a [homoglyph](https://en.wikipedia.org/wiki/Homoglyph) might be used to trick a person into using the wrong asset code.
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:122]:
```suggestion
While it is possible to store arbitrary data in this field, implementations that detect the above recommended character conformance should display them as human-readable text, allowing issuers to support well-known ISO-4207 currency codes in addition to custom codes. This also helps prevents spoofing attacks where a [homoglyph](https://en.wikipedia.org/wiki/Homoglyph) might be used to trick a person into using the wrong asset code.
```

File: XLS-0033d-CFTs/README.md [hunk 10]
```diff
@@ -0,0 +1,836 @@
...
+A **`CFTokenPage`** object may have the following required and optional fields:
+
+| Field Name          | Required? | JSON Type | Internal Type |
+| ------------------- |-----------| --------- |---------------|
+| `LedgerEntryType`   | ✔️         | `string`  | `UINT16`      |
+| `PreviousPageMin`   | ️          | `string`  | `UINT256`     |
+| `NextPageMin`       | ️          | `string`  | `UINT256`     |
+| `PreviousTxnID`     | ️          | `string`  | `HASH256`     |
+| `PreviousTxnLgrSeq` | ️          | `number`  | `UINT32`      |
+| `CFTokens`          | ️ ✔        | `object`  | `TOKEN`        |
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:244]:
```suggestion
| `LedgerEntryType`   | ✔️         | `string`  | `UINT16`      |
| `PreviousPageMin`   | ️          | `string`  | `UINT256`     |
| `NextPageMin`       | ️          | `string`  | `UINT256`     |
| `PreviousTxnID`     | ️          | `string`  | `HASH256`     |
| `PreviousTxnLgrSeq` | ️          | `number`  | `UINT32`      |
| `CFTokens`          | ️✔        | `object`  | `TOKEN`        |
```

File: XLS-0033d-CFTs/README.md [hunk 11]
```diff
@@ -0,0 +1,836 @@
...
+
+| Field Name      | Required?          | JSON Type | Internal Type |
+| --------------- | ------------------ | --------- | ------------- |
+| `MaximumAmount` | :heavy_check_mark: | `string`  | `UINT64`      | 
+
+The maximum asset amount of this token that should ever be issued.
+
+| Field Name | Required?          | JSON Type | Internal Type |
+| ---------- | ------------------ | --------- | ------------- |
+| `CFTMetadata` | :heavy_check_mark: | `string`  | `BLOB`        |
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:395]:
```suggestion
| Field Name    | Required?          | JSON Type | Internal Type |
| ------------- | ------------------ | --------- | ------------- |
| `CFTMetadata` | :heavy_check_mark: | `string`  | `BLOB`        | 
```

File: XLS-0033d-CFTs/README.md [hunk 12]
```diff
@@ -0,0 +1,836 @@
...
+--------------------------------
+LowQualityIn                  32
+LowQualityOut                 32
+HighQualityIn                 32
+HighQualityOut                32
+--------------------------------
+MAXIMUM TOTAL SIZE:         2000 (250 bytes)
+```
+
+TODO: Validate these numbers as they may be slightly low for truslintes. For example, in addition to the above data, trustlines require two directory entries for low/high nodes that get created for each trustline (i.e., for each RippleState object). This creates two root objects, each 98 bytes, adding 196 bytes in total per unique issuer/holder. Conversely, for CFTs, the page structure allows for up to 32 issuances to be held by a single token holder while only incurring around 102 bytes for a CFTokenPage. Thus, every time an account wants to hold a new token, the current Trustline implementation would require _at least_ 430 bytes every time. If we imagine a single account holding 20 tokens, CFTs would require ~1040 bytes, whereas trustlines would require ~8,600 bytes!
```
Comment (intelliot) [inline XLS-0033d-CFTs/README.md:829]:
```suggestion
TODO: Validate these numbers as they may be slightly low for trust lines. For example, in addition to the above data, trust lines require two directory entries for low/high nodes that get created for each trust line (i.e. for each RippleState object). This creates two root objects, each 98 bytes, adding 196 bytes in total per unique issuer/holder. Conversely, for CFTs, the page structure allows for up to 32 issuances to be held by a single token holder while only incurring around 102 bytes for a CFTokenPage. Thus, every time an account wants to hold a new token, the current trust line implementation would require _at least_ 430 bytes. If we imagine a single account holding 20 tokens, CFTs would require ~1040 bytes, whereas trust lines would require ~8,600 bytes!
```

**Review by intelliot — APPROVED** [review]

**Comment by sappenin** [comment]
+1 to all of the changes proposed by @intelliot (I can't merge them though).

:)

File: XLS-0033d-CFTs/README.md [hunk 13]
```diff
@@ -0,0 +1,836 @@
+<pre>
+Title:       <b>Compact Fungible Tokens (CFTs)</b>
+Revision:    <b>1</b> (2022-08-05)
```
[the remark this answers is not in the document]
Comment (sappenin) [reply XLS-0033d-CFTs/README.md:3]: I guess technically this is revision `2`, FWIW.

File: XLS-0033d-CFTs/README.md [hunk 14]
```diff
@@ -0,0 +1,836 @@
...
+| ------------- | ------------------ | --------- | ------------- |
+| `CFTMetadata` | :heavy_check_mark: | `string`  | `BLOB`        | 
+
+Arbitrary metadata about this issuance, in hex format. The limit for this field is 1024 bytes.
+
+##### 1.3.1.1.2 Example **`CFTokenIssuanceCreate`** transaction
+
+```
+{
+  "TransactionType": "CFTokenIssuanceCreate",
```
Comment (sappenin) [inline XLS-0033d-CFTs/README.md:403]:
```suggestion
  "Account": "rajgkBmMxmz161r8bWYH7CQAFZP5bA9oSG",
  "TransactionType": "CFTokenIssuanceCreate",
```

File: XLS-0033d-CFTs/README.md [hunk 15]
```diff
@@ -0,0 +1,836 @@
...
+
+If the transaction is successful, the newly created token will be owned by the account (the `creator` account) which executed the transaction.
+
+##### 1.3.1.1.1 Transaction-specific Fields
+| Field Name         | Required? | JSON Type | Internal Type |
+| ------------------ | --------- | --------- |---------------|
+| `TransactionType`  | ️ ✔        | `object`  | `UINT16`      |
+
+Indicates the new transaction type **`CFTokenIssuanceCreate`**. The integer value is `25 (TODO)`.
+
```
Comment (sappenin) [inline XLS-0033d-CFTs/README.md:351]:
```suggestion

| Field Name         | Required? | JSON Type | Internal Type |
| ------------------ | --------- | --------- |---------------|
| `Account`              | ️ ✔          | `string`   | `ACCOUNT`    |

The account to create the issuance under.
```

File: XLS-0033d-CFTs/README.md [hunk 16]
```diff
@@ -0,0 +1,836 @@
...
+Identifies the page in the owner's directory where this item is referenced.
+
+##### 1.2.1.1.2. Example **`CFTokenIssuance`** JSON
+
+ ```json
+ {
+     "LedgerEntryType": "CFTokenIssuance",
+     "Flags" : 131072,
+     "Issuer" : "rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW",
+     "AssetCode": "5553440000000000000000000000000000000000",
```
Comment (sappenin) [inline XLS-0033d-CFTs/README.md:151]:
```suggestion
     "AssetCode": "0000000000000000000000005553440000000000",
```

File: XLS-0033d-CFTs/README.md [hunk 17]
```diff
@@ -0,0 +1,836 @@
+<pre>
+Title:       <b>Compact Fungible Tokens (CFTs)</b>
+Revision:    <b>1</b> (2023-05-22)
```
Comment (Silkjaer) [inline XLS-0033d-CFTs/README.md:3]: Revision should be removed

File: XLS-0033d-CFTs/README.md [hunk 18]
```diff
@@ -0,0 +1,836 @@
+<pre>
```
Comment (Silkjaer) [inline XLS-0033d-CFTs/README.md:1]: Folder name should have "CFTs" in lowercase

File: XLS-0033d-CFTs/README.md [hunk 19]
```diff
@@ -0,0 +1,836 @@
...
+Title:       <b>Compact Fungible Tokens (CFTs)</b>
+Revision:    <b>1</b> (2023-05-22)
+
```
Comment (Silkjaer) [inline XLS-0033d-CFTs/README.md:4]: Type: draft should be added

**Review by Silkjaer — COMMENTED** [review]
A few fixes needed before it can be merged :)

File: XLS-0033d-CFTs/README.md [hunk 20]
```diff
@@ -0,0 +1,836 @@
...
+#### 1.2.1.2. The **`CFToken`** object
+The **`CFToken`** object represents an amount of a token held by an account that is **not** the token issuer. CFTs are acquired via ordinary Payment or DEX transactions, and can optionally be redeemed or exchanged using these same types of transactions.
+
+##### 1.2.1.2.1 Fields
+A **`CFToken`** object can have the following required and optional fields. Notice that, unlike other objects, no field is needed to identify the object type or current owner of the object, because CFT holdings are grouped into pages that implicitly define the object type and identify the holder.
+
+| Field Name            | Required?          | JSON Type | Internal Type |
+| --------------------- |--------------------|-----------|---------------|
+| `CFTIssuanceID`       | :heavy_check_mark: |  `string` | `UINT256`     |
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:189]: should we call it `CFTokenIssuanceID` to make it consistent with rest of the naming?

Comment (sappenin) [reply XLS-0033d-CFTs/README.md:189]: I think that's a good idea. @ledhed2222 can you make that change?

File: XLS-0033d-CFTs/README.md [hunk 21]
```diff
@@ -0,0 +1,836 @@
...
+| `LedgerEntryType`   | :heavy_check_mark: | `number`  | `UINT16`      |
+| `Flags`             | :heavy_check_mark: | `number`  | `UINT32`      |
+| `Issuer`            | :heavy_check_mark: | `string`  | `ACCOUNTID`   |
+| `AssetCode`         | :heavy_check_mark: | `string`  | `UINT160`     |
+| `AssetScale`        | (default)          | `number`  | `UINT8`       |
+| `MaximumAmount`     | :heavy_check_mark: | `string`  | `UINT64`      |
+| `OutstandingAmount` | :heavy_check_mark: | `string`  | `UINT64`      |
+| `LockedAmount`      | ️(default)          | `string`  | `UINT64`      |
+| `TransferFee`       | ️(default)          | `number`  | `UINT16`      |
+| `CFTMetadata`       |                    | `string`  | `BLOB`        |
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:92]: why do we have to explicitly name it `CFTMetadata`? could it just be called `Metadata`?

File: XLS-0033d-CFTs/README.md [hunk 22]
```diff
@@ -0,0 +1,836 @@
...
+ }
+ ```
+ 
+#### 1.3.3.1.1 CFTokenSet Flags
+Transactions of the `CFTokenFreeze` type support additional values in the Flags field, as follows:
+
+| Flag Name         | Flag Value | Description |
+|-------------------|------------|-------------|
+| `tfSetFreeze`     | ️`0x0001`  | If set, indicates that all CFT balances for this asset should be frozen. |
+| `tfSetUnFreeze`   | ️`0x0002`  | If set, indicates that all CFT balances for this asset should be unfrozen. |
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:488]: `tfSetUnFreeze` is supposed to be called `tfClearFreeze` instead, if the plan is to re-use `TrustSet` flag constants
```
constexpr std::uint32_t tfSetFreeze                        = 0x00100000;
constexpr std::uint32_t tfClearFreeze                      = 0x00200000;
```

File: XLS-0033d-CFTs/README.md [hunk 23]
```diff
@@ -0,0 +1,836 @@
...
+
+###### 1.2.1.2.1.1. `CFTIssuanceID`
+
+The `CFTokenIssuance` identifier.
+
+###### 1.2.1.2.1.2. `Amount`
+
+This value specifies a positive amount of tokens currently held by the owner. Valid values for this field are between 0x0 and 0xFFFFFFFFFFFFFFFF.
+
+###### 1.2.1.2.1.3. `LockedAmount`
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:202]: is there any use for `LockedAmount` field is the phase 1 implementation?

File: XLS-0033d-CFTs/README.md [hunk 24]
```diff
@@ -0,0 +1,836 @@
...
+
+| Field Name        | Required? | JSON Type | Internal Type |
+| ----------        | --------- | --------- | ------------- |
+| `TransactionType` |  ✔️        | `string`  | `UINT16`      | 
+
+Indicates the new transaction type **`CFTokenIssuanceDestroy`**. The integer value is `26` (TODO).
+
+| Field Name  | Required? | JSON Type | Internal Type |
+| ----------- | --------- | --------- | ------------- |
+| `CFTokenIssuance` |  ✔️        | `string`  | `UINT256`     |
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:432]: should it say `CFTokenIssuanceID`?

File: XLS-0033d-CFTs/README.md [hunk 25]
```diff
@@ -0,0 +1,836 @@
...
+#### 1.3.3.1 CFTokenIssuanceSet
+| Field Name         | Required? | JSON Type | Internal Type |
+| ------------------ | --------- | --------- |---------------|
+| `TransactionType`  | ️ ✔        | `object`  | `UINT16`      |
+
+Indicates the new transaction type **`CFTokenIssuanceSet`**. The integer value is `28 (TODO)`.
+
+| Field Name  | Required? | JSON Type | Internal Type |
+| ----------- | --------- | --------- | ------------- |
+| `CFTokenIssuance` |  ✔️  | `string`  | `UINT256`     |
```
Comment (shawnxie999) [inline XLS-0033d-CFTs/README.md:457]: ditto

File: XLS-0033d-CFTs/README.md (hunk 23 above)
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:202]: nevermind, this is used for freeze.

File: XLS-0033d-CFTs/README.md (hunk 21 above)
Comment (sappenin) [reply XLS-0033d-CFTs/README.md:92]: I might be misunderstanding the existing code, but `Metadata` is [already defined](https://github.com/XRPLF/rippled/blob/adf672ff83d473043485f8a48d39bb5ecbd892b5/src/ripple/protocol/impl/SField.cpp#L81) as `UNTYPED`. In our case, we want this to be typed as a `BLOB` or `VL` (I think) -- similar to [sfMemoData](https://github.com/XRPLF/rippled/blob/adf672ff83d473043485f8a48d39bb5ecbd892b5/src/ripple/protocol/impl/SField.cpp#L261) (for example).

Comment (sappenin) [reply XLS-0033d-CFTs/README.md:92]: > But should we rename it to CFTokenMetadata?

Yes, I think so (we should be consistent with that naming).

**Comment by intelliot** [comment]
@ledhed2222 please respond to the above comments, at your convenience

File: XLS-0033d-CFTs/README.md [hunk 26]
```diff
@@ -0,0 +1,836 @@
...
+}
+```
+
+This transaction assumes that the issuer of the token is the signer of the transaction.
+
+### 1.3.2 The **`CFTokenIssuanceDestroy`** transaction
+
+The **`CFTokenIssuanceDestroy`** transaction is used to remove an **`CFTokenIssuance`** object from the directory node in which it is being held, effectively removing the token from the ledger ("destroying" it).
+
+If this operation succeeds, the corresponding **`CFTokenIssuance`** is removed and the owner’s reserve requirement is reduced by one.
```
[the remark this answers is not in the document]
Comment (ledhed2222) [reply XLS-0033d-CFTs/README.md:420]: This is already described in the section on `AccountDelete` - you must destroy any CFTIssuances you have before you can delete your account. Added a statement to the `CFTokenIssuanceDestroy` tx to indicate that you cannot destroy a CFTIssuance if there are any existing holders.

**Comment by ledhed2222** [comment]
@intelliot done :)

**Review by intelliot — APPROVED** [review]

**Review by shawnxie999 — APPROVED** [review]

**Review by sappenin — APPROVED** [review]

**Comment by ledhed2222** [comment]
@Silkjaer hey there! can you review again please?

**Review by Silkjaer — APPROVED** [review]
LGTM!

**Pull request merged**